### ISSUE TITLE
Question concerning softmax for ava-style detection tasks

### ISSUE THREAD

**Issue by FabiTheGabi** [facebookresearch/SlowFast#261]
Dear Slowfast-Team,

I have designed a custom dataloader for a custom dataset in the ava style.
Currently, I am able to train models on custom datasets using the standard AVA scripts provided in the repository.

My new dataset's action categories are mutually exclusive, and thus I would like to use "softmax" as HEAD_ACT and "cross_entropy" as LOSS_FUNC.

This seems to be not supported so far and in issue https://github.com/facebookresearch/SlowFast/pull/208 it is (in my opinion correctly) pointed out that this line of code should be "self.act = nn.Softmax(dim=1)" https://github.com/facebookresearch/SlowFast/blob/0c3df9dc3884ade5a9a2ee14fd3cd5eab3883a7a/slowfast/models/head_helper.py#L94

Furthermore, I adjusted the input labels and they match the Kinetics-style, which allows to compute the loss https://github.com/facebookresearch/SlowFast/blob/0c3df9dc3884ade5a9a2ee14fd3cd5eab3883a7a/tools/train_net.py#L78

But my model does not significantly improve during training. 

It would be awesome if you could give me hint, what I did not think of so far. 
Here is my config file I use for training:
[Softmax_SLOW_8x8_R50_SHORT .txt](https://github.com/facebookresearch/SlowFast/files/4996604/Softmax_SLOW_8x8_R50_SHORT.txt)

Thank you a lot in advance for your support and your awesome repository.

Best regards.

**Comment by FabiTheGabi** [comment]
Hi all,

I tried to solve this problem myself:

The following steps are necessary:
0. Set MODEL.LOSS_FUNC = cross_entropy and MODEL.HEAD_ACT = "softmax"
1. It is necessary to change "self.act = nn.Softmax(dim=1)" according to my question above.
2. Create the labels in the required format for https://pytorch.org/docs/master/generated/torch.nn.CrossEntropyLoss.html
3. Since the loss function combines nn.LogSoftmax() and nn.NLLLoss() in one single class, make sure that in the training process, you feed the model output without applying the softmax function to the loss_function to get the correct loss  
Last: Try out the best lr_policy ;)

Best regards

**Comment by mikeyEcology** [comment]
Hi @FabiTheGabi , I'm trying to do something similar where I want to create a custom dataset in the vein of AVA. Would you mind sharing your code of how you set up the loader? Did you go through the registration steps ([described here](https://detectron2.readthedocs.io/tutorials/datasets.html))? Or did you manually set up your dataset?
Thank you

**Comment by FabiTheGabi** [comment]
Hi @mikeyEcology, you can see my modifications in [my VideoProcessMining repository](https://github.com/FabiTheGabi/VideoProcessMining), where I also provide a custom data loader and describe how the data preparation is done. It works for all data in the AVA format (spatio temporal activity recognition).
I hope it helps you. 
Best regards

**Comment by mikeyEcology** [comment]
Thank you @FabiTheGabi !

**Comment by Staniwnl** [comment]
@FabiTheGabi Hi, Thank you so much for the data loader repo. Can I ask you a question here? What is the purpose of the "ava_train_predicted_boxes.csv" file?  
And when I tried to train the AVA dataset, I followed the recipe in [https://github.com/facebookresearch/SlowFast/blob/master/configs/AVA/SLOWFAST_32x2_R50_SHORT.yaml](url)
At the time of val, slowfast seems to only use "ava_detection_val_boxes_and_labels.csv" as the ground truth file. Is this setting correct?

**Comment by FabiTheGabi** [comment]
Hi @Staniwnl,
sorry for the late response! The ava_train_predicted_boxes.csv contains additional bounding boxes as training input that have a high IoU with the train_groundtruth bounding boxes. The additional training input kind of avoids overfitting to the given bounding boxes and helps your learner to get results results based on the bounding boxes determined by your object prediction model. The high IoU is important to be able to label these bounding boxes with respect to the original activity classes in your dataset.
Thus, it is not necessary to include any information contained in this file in your val setting, since it does not contain any information on your val dataset. 

Hope this helps. :)

**Comment by Staniwnl** [comment]
@FabiTheGabi  Thank you very much for clearing that up, it helped me a lot.
